pkg/binary/envoy/debug/pii.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"bitbucket.org/creachadair/shell"
+)
+
+// defaultHash returns the hashed value of s using sha256 defaultHash function
+// TODO: salt the Hash
+//nolint: unused,deadcode
+func defaultHash(s string) string {
+	h := sha256.New()
+	_, _ = h.Write([]byte(s))
+	return string(h.Sum(nil))
+}
+
+var defaultFormat = `[%START_TIME%] "%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%"` +
+	` %RESPONSE_CODE% %RESPONSE_FLAGS% "%DYNAMIC_METADATA(istio.mixer:status)%" "%REQ(USER-AGENT)%"`
+var defaultPII = map[string]bool{
+	"[%START_TIME%]": true,
+}
+
+// Filter filters log fields using pii and modify all PII fields using hash
+type Filter struct {
+	hash   func(string) string
+	pii    map[string]bool
+	format []string
+}
+
+// NewFilter constructs a custom filter object
+func NewFilter(formatStr string, hash func(string) string, pii map[string]bool) (Filter, error) {
```

Comment:
You haven't had to deal with go packages yet so you haven't been trained to think in terms of visibility, but at the moment this interface is not usable as you intend. Primarily, there's no way for a user to access `defaultFormat`, `defaultHash`, or `defaultPII` outside of this package, so no user calling it would be able to instantiate a default Filter.

Further, the only times we don't use the default formatStr, hash, and PII are in tests. Our tests are in the same package so they can just construct a Filter directly themselves with the correct fields. The default.

So two options: 1) for now make this constructor take no arguments and return a new Filter with the default values - callers can't customize at all (less extreme version: use default hash and PII, but accept the format string from the caller; make the default format string a public variable of the package so callers can use it). 2) Add a second constructor:

```go
func New() (Filter, error) {
	return NewFilter(defaultFormat, defaultHash, defaultPII)
}
```

Either way we need to give people outside of this package a way to get an instance that has the default values.